### ISSUE TITLE
Deployments can fail if artifact retrieval fails due to service accounts not having full control permissions

### ISSUE THREAD

**Issue by jburger** [OctopusDeploy/Issues#5433]
# Prerequisites

- [x] I have verified the problem exists in the latest version
- [x] I have searched [open](https://github.com/OctopusDeploy/Issues/issues) and [closed](https://github.com/OctopusDeploy/Issues/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed) issues to make sure it isn't already reported
- [x] I have written a descriptive issue title
- [x] I have linked the original source of this report
- [x] I have tagged the issue appropriately (area/*, kind/bug, tag/regression?)

# The bug

In limited access scenarios, an Octopus Server account with read/write permissions over its artifact directory may fail when uploading artifacts during a deployment if `full control` is not granted over the artifact directory. Despite this failure, the artifact is written to the directory.

This occurs because the artifact upload process attempts to proactively adjust the artifact ACL to prevent [this bug](https://github.com/OctopusDeploy/Issues/issues/3881) from happening _after_ the artifact is written. 

Full control allows a user to adjust the permissions of a file after it has been written. In this failure scenario, the artifacts inherit the permissions of the folder they are in and the service account is only granted write permission, so is unable to change the file permissions here. 

## What I expected to happen

While [our documentation](https://octopus.com/docs/installation/permissions-for-the-octopus-windows-service) does explain that `full control` is required over the Home directories (which includes artifact directories), a failure to adjust these ACLs shouldn't create a failed deployment, since the artifacts are written and accessible in other circumstances.

## Steps to reproduce

1. Configure a test environment to revoke `change permissions` and `take ownership` over the artifact folder for the Octopus Server service account.
![image](https://user-images.githubusercontent.com/579657/55214104-9787ef00-5245-11e9-879e-4b18aa20cf2b.png)

2. Deploy a project that makes use of `New-OctopusArtifact`. NOTE: this must be executed on a tentacle and not on the server.
3. The deployment should fail with the following log excerpt.

```
Collecting artifact [REDACTED]
08:40:13 Verbose | Attempted to perform an unauthorized operation.
| System.UnauthorizedAccessException
| at System.Security.AccessControl.Win32.SetSecurityInfo(ResourceType type, String name, SafeHandle handle, SecurityInfos securityInformation, SecurityIdentifier owner, SecurityIdentifier group, GenericAcl sacl, GenericAcl dacl)
| at System.Security.AccessControl.NativeObjectSecurity.Persist(String name, SafeHandle handle, AccessControlSections includeSections, Object exceptionContext)
| at System.Security.AccessControl.NativeObjectSecurity.Persist(String name, AccessControlSections includeSections, Object exceptionContext)
| at System.Security.AccessControl.FileSystemSecurity.Persist(String fullPath)
| at Halibut.Transport.Protocol.TemporaryFileStream.SaveTo(String filePath) in TemporaryFileStream.cs:line 24
| at Octopus.Server.Orchestration.Targets.Artifacts.ArtifactStore.Save(String name, DataStream dataStream, String logCorrelationId)
| at Octopus.Server.Orchestration.Targets.Tentacles.TentacleRemoteEndpointFacade.CollectArtifacts(IEnumerable`1 artifacts)
| at Octopus.Server.Orchestration.Targets.Tentacles.ScriptObserver.ObserveThenFinish(ScriptTicket ticket, ScriptOutputFilter output, Action`1 collectArtifacts)
| at Octopus.Server.Orchestration.Targets.Tentacles.ScriptObserver.ObserveUntilComplete(ScriptTicket ticket, Action`1 collectArtifacts)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.ActionExecution.Immediate.ExecutionTargets.TentacleExecutionTarget.Execute(String bootstrapper, IReadOnlyList`1 boostrapperArguments, IReadOnlyList`1 files, Nullable`1 forceIsolationLevel, Boolean raw)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.ActionExecution.Immediate.ImmediateExecutor.ExecuteCalamari(ICalamariFlavour flavour, String calamariCommand, IReadOnlyList`1 calamariArguments, IReadOnlyList`1 files, IReadOnlyList`1 deploymentTools, VariableCollection extraVariables, TargetManifest targetManifest)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.ActionExecution.CommandBuilders.RunCalamariCommandBuilder.Execute()
| at Octopus.Server.Orchestration.ServerTasks.Deploy.Steps.Script.ScriptActionHandler.Execute(IActionHanderContext context)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.ActionDispatch.NewActionDispatcher.<>c__DisplayClass12_0.<Execute>b__2()
| at Octopus.Server.Orchestration.ServerTasks.Deploy.ActionDispatch.NewActionDispatcher.ExecuteOnDeploymentTarget(ActionCommand command, DeploymentTarget deploymentTarget, Maybe`1 guidanceExcludeCallback)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.DeploymentTaskController.<>c__DisplayClass36_1.<ExecuteActionAndInitLoggingContext>b__1()
| at Octopus.Server.Orchestration.ServerTasks.Deploy.DeploymentTaskController.<>c__DisplayClass36_0.<ExecuteActionAndInitLoggingContext>b__0()
| at Octopus.Server.Orchestration.ServerTasks.Deploy.DeploymentTaskController.ExecuteWithTransientErrorDetection(Action action, DeploymentTarget deploymentTarget)
| at Octopus.Server.Orchestration.ServerTasks.Deploy.DeploymentTaskController.ExecuteActionAndInitLoggingContext(PlannedStep step, DeploymentTarget targetContext, PlannedAction action)
|
| Canceled: Apply retention policy on Tentacles
08:40:14 Verbose | Step "Apply Tentacle Retention Policy" runs only when all previous steps succeeded; skipping
| 
```

## Affected versions

**Octopus Server: 2018.9.7 and later**
**Halibut: 4.0.5 and later**

## Workarounds

Relax the folder permissions to allow the service full control over the artifacts directory.

## Links
[Internal support ticket](https://secure.helpscout.net/conversation/810034314/41293?folderId=2118197)

**Comment by jburger** [comment]
My bad, this will be fixed once this PR lands: https://github.com/OctopusDeploy/OctopusDeploy/pull/3653

**Comment by jburger** [comment]
Fixed in Octopus Server `2019.4.1` and also released in LTS `2018.10.7` and LTS `2019.3.2`
